File: javascript/ch6/computeAverageGrade.js
const R = require('ramda');

const fork = (join, func1, func2) => (value) => join(func1(value), func2(value));

const toLetterGrade = (grade) => {
  if (grade >= 90) {
    return 'A';
  }
  if (grade >= 80) {
    return 'B';
  }
  if (grade >= 70) {
    return 'C';
  }
  if (grade >= 60) {
    return 'D';
  }
  return 'F';
}

const computeAverageGrade = R.compose(toLetterGrade, fork(R.divide, R.sum, R.length));

module.exports = {
  fork,
  toLetterGrade,
  computeAverageGrade
};

File: python/python/ch6/compute_average_grade.py
import pydash
import toolz as tz

def fork(join, func1, func2): 
    def inner_fork(value):
        return join(func1(value), func2(value))
    return inner_fork

def to_letter_grade(grade):
    if grade >= 90:
        return 'A'
    if grade >= 80:
        return 'B'
    if grade >= 70:
        return 'C'
    if grade >= 60:
        return 'D'
    return 'F'
  
def compute_average_grade(arr):
    compose_func = tz.compose(
        to_letter_grade,
        fork(
            pydash.divide, 
            pydash.sum_, 
            pydash.size
        )
    )
    return compose_func(arr)
